### PR TITLE
Symbol cleanup

### PR DESCRIPTION
**Pull request by aehyvari**
This PR cleans further the symbol interface by forbidding (incorrect) changes to symbols once created.

### CODE REVIEW

**Comment by blishko** [comment]
This looks very nice! Can you rebase so my commits from the previous PR don't get mixed into the diff?

File: src/logics/Logic.cc [hunk 1]
```diff
@@ -717,7 +717,7 @@ PTRef Logic::mkConst(const char* name)
 
 
 PTRef Logic::mkVar(SRef s, const char* name) {
-    SymRef sr = newSymb(name, {s});
+    SymRef sr = sym_store.newSymb(name, {s}, SymConf::Default);
```
Comment (blishko) [inline src/logics/Logic.cc:720]:
```suggestion
    SymRef sr = sym_store.newSymb(name, {s});
```

File: src/logics/Logic.cc [hunk 2]
```diff
@@ -80,7 +80,7 @@ Logic::Logic(opensmt::Logic_t _logicType) :
     , term_FALSE(mkConst(getSort_bool(), tk_false))
     , sym_TRUE(getSymRef(term_TRUE))
     , sym_FALSE(getSymRef(term_FALSE))
-    , sym_ANON(sym_store.newSymb(tk_anon, {}))
+    , sym_ANON(sym_store.newSymb(tk_anon, {}, SymConf::Default))
```
Comment (blishko) [inline src/logics/Logic.cc:83]:
```suggestion
    , sym_ANON(sym_store.newSymb(tk_anon, {}))
```

File: src/logics/Logic.cc [hunk 3]
```diff
@@ -810,19 +806,19 @@ void Logic::instantiateFunctions(SRef sr)
     sortToIte.insert(sr, tr);
 }
 
-SymRef Logic::declareFun(const char* fname, const SRef rsort, const vec<SRef>& args, char** msg, bool interpreted)
+SymRef Logic::declareFun(const char* fname, const SRef rsort, const vec<SRef>& args, SymbolConfig const & symbolConfig, char** msg)
```
Comment (blishko) [inline src/logics/Logic.cc:809]:
```suggestion
SymRef Logic::declareFun(const char* fname, const SRef rsort, const vec<SRef>& args, SymbolConfig const & symbolConfig)
```

**Review by blishko — COMMENTED** [review]
Looks very good! I think we can get rid of the overload of `declareFun` with `char**` and also take advantage of the fact that there is an overload which passes the `Default` symbol config.

File: src/symbols/Symbol.h [hunk 4]
```diff
@@ -36,35 +36,69 @@ WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
 typedef uint32_t SymId; // Used as an array index
 
+enum class SymbolProperty : int {
```
[the remark this answers is not in the document]
Comment (aehyvari) [reply src/symbols/Symbol.h:39]: Yes.

File: src/bin/Interpret.cc [hunk 5]
```diff
@@ -803,7 +803,7 @@ bool Interpret::declareFun(ASTNode const & n) // (const char* fname, const vec<S
     for (int i = 1; i < args.size(); i++)
         args2.push(args[i]);
 
-    SymRef rval = logic->declareFun(fname, rsort, args2, &msg);
+    SymRef rval = logic->declareFun(fname, rsort, args2, SymConf::Default, &msg);
```
[the remark this answers is not in the document]
Comment (aehyvari) [reply src/bin/Interpret.cc]: Fixed.

File: src/logics/Logic.h [hunk 6]
```diff
@@ -242,21 +240,22 @@ class Logic {
...
-
-    SymRef      declareFun_LeftAssoc(std::string const & s, SRef rsort, vec<SRef> const & args) { SymRef sr = declareFun(s, rsort, args); sym_store[sr].setLeftAssoc(); return sr; }
-    SymRef      declareFun_RightAssoc(std::string const & s, SRef rsort, vec<SRef> const & args) { SymRef sr = declareFun(s, rsort, args); sym_store[sr].setRightAssoc(); return sr; }
-    SymRef      declareFun_Chainable(std::string const & s, SRef rsort, vec<SRef> const & args) { SymRef sr = declareFun(s, rsort, args); sym_store[sr].setChainable(); return sr; }
-    SymRef      declareFun_Pairwise(std::string const & s, SRef rsort, vec<SRef> const & args) { SymRef sr = declareFun(s, rsort, args); sym_store[sr].setPairwise(); return sr;}
+    SymRef      declareFun(const char* fname, const SRef rsort, const vec<SRef>& args, SymbolConfig const & symbolConfig, char** msg);
+    SymRef      declareFun(const std::string & fname, const SRef rsort, const vec<SRef>& args, SymbolConfig const & symbolConfig) { char *msg; return declareFun(fname.data(), rsort, args, symbolConfig, &msg); };
```
[the remark this answers is not in the document]
Comment (aehyvari) [reply src/logics/Logic.h]: Should be fixed now.

File: src/logics/Logic.cc (hunk 3 above)
Comment (aehyvari) [reply src/logics/Logic.cc]: Fixed.

**Review by blishko — APPROVED** [review]
LGTM!

**Pull request merged**